lib/Sema/ConstraintSystem.cpp

```diff
@@ -2729,6 +2729,154 @@ static void diagnoseOperatorAmbiguity(ConstraintSystem &cs,
               llvm::join(parameters, ", "));
 }
 
+std::string swift::describeGenericType(ValueDecl *GP, bool includeName) {
+  if (!GP)
+    return "";
+
+  Decl *parent = nullptr;
+  if (auto *AT = dyn_cast<AssociatedTypeDecl>(GP)) {
+    parent = AT->getProtocol();
+  } else {
+    auto *dc = GP->getDeclContext();
+    parent = dc->getInnermostDeclarationDeclContext();
+  }
+
+  if (!parent)
+    return "";
+
+  llvm::SmallString<64> result;
+  llvm::raw_svector_ostream OS(result);
+
+  OS << Decl::getDescriptiveKindName(GP->getDescriptiveKind());
+
+  if (includeName && GP->hasName())
+    OS << " '" << GP->getBaseName() << "'";
+
+  OS << " of ";
+  OS << Decl::getDescriptiveKindName(parent->getDescriptiveKind());
+  if (auto *decl = dyn_cast<ValueDecl>(parent)) {
+    if (decl->hasName())
+      OS << " '" << decl->getFullName() << "'";
+  }
+
+  return OS.str();
+}
+
+/// Special handling of conflicts associated with generic arguments.
+///
+/// func foo<T>(_: T, _: T) {}
+/// func bar(x: Int, y: Float) {
+///   foo(x, y)
+/// }
+///
+/// It's done by first retrieving all generic parameters from each solution,
+/// filtering boundings into distrinct set and diagnosing any differences.
+static bool diagnoseConflictingGenericArguments(ConstraintSystem &cs,
+                                                const SolutionDiff &diff,
+                                                ArrayRef<Solution> solutions) {
+  if (!diff.overloads.empty())
+    return false;
+
+  if (!llvm::all_of(solutions, [](const Solution &solution) -> bool {
+        return llvm::all_of(
+            solution.Fixes, [](const ConstraintFix *fix) -> bool {
+              return fix->getKind() == FixKind::AllowArgumentTypeMismatch ||
+                     fix->getKind() == FixKind::AllowFunctionTypeMismatch ||
+                     fix->getKind() == FixKind::AllowTupleTypeMismatch;
+            });
+      }))
+    return false;
+
+  auto &DE = cs.getASTContext().Diags;
+
+  llvm::SmallDenseMap<TypeVariableType *, SmallVector<Type, 4>> conflicts;
+
+  for (const auto &binding : solutions[0].typeBindings) {
+    auto *typeVar = binding.first;
+
+    if (!typeVar->getImpl().getGenericParameter())
+      continue;
+
+    llvm::SmallSetVector<Type, 4> arguments;
+    arguments.insert(binding.second);
+
+    if (!llvm::all_of(solutions.slice(1), [&](const Solution &solution) {
+          auto binding = solution.typeBindings.find(typeVar);
+          if (binding == solution.typeBindings.end())
+            return false;
+
+          if (auto *opaque =
+                  binding->second->getAs<OpaqueTypeArchetypeType>()) {
+            auto *decl = opaque->getDecl();
+            arguments.remove_if([&](Type argType) -> bool {
```

Comment:
Could you add a short comment for why this is necessary? Took me a second to figure it out!

Reply:
Done!